cpp/src/sampling/random_walks_impl.cuh

```diff
@@ -174,16 +352,203 @@ struct node2vec_selector {
     GraphViewType const& graph_view,
     std::optional<edge_property_view_t<typename GraphViewType::edge_type, weight_t const*>>
       edge_weight_view,
-    rmm::device_uvector<typename GraphViewType::vertex_type> const& current_vertices)
+    rmm::device_uvector<typename GraphViewType::vertex_type>& current_vertices,
+    std::optional<rmm::device_uvector<typename GraphViewType::vertex_type>>& previous_vertices)
   {
     //  To do node2vec, I need the following:
-    //    1) transform_reduce_dst_nbr_intersection_of_e_endpoints_by_v to compute the sum of the
-    //       node2vec style weights
-    //    2) Generate a random number between [0, output_from_trdnioeebv[v])
-    //    3) a sampling value that lets me pick the correct edge based on the same computation
-    //       (essentially weighted sampling, but with a function that computes the weight rather
-    //       than just using the edge weights)
-    CUGRAPH_FAIL("node2vec not implemented");
+    //    1) Find nbr_intersection() CSR
+    //    2) Convert CSR into COO format, with (p,c,d) triples
+    //    3) Shuffle COO structure s.t. each entry is on GPU where (c,d) resides
+    //    4) Revert COO back to CSR
+    //  After this, existing functor should work for symmetric graphs.  Add CUGRAPH_EXPECTS to ensure
+    //  no asymmetric graphs and add FIXME
+
+    //  Create vertex frontier
+    using vertex_t = typename GraphViewType::vertex_type;
+    
+    using tag_t = vertex_t;
+
+    cugraph::vertex_frontier_t<vertex_t, tag_t, GraphViewType::is_multi_gpu, false> vertex_frontier(handle, 1);
+    vertex_frontier.bucket(0).insert(thrust::make_zip_iterator(current_vertices.begin(),
+                                                              (*previous_vertices).begin()), 
+                                     thrust::make_zip_iterator(current_vertices.end(), 
+                                                              (*previous_vertices).end()));         
+
+    //  Zip previous and current vertices for nbr_intersection()
+    auto intersection_pairs = thrust::make_zip_iterator(current_vertices.begin(), (*previous_vertices).begin());
+    
+    auto [intersection_offsets, intersection_indices] =
+      detail::nbr_intersection(handle,
+                               graph_view,
+                               cugraph::edge_dummy_property_t{}.view(),
+                               intersection_pairs,
+                               intersection_pairs + current_vertices.size(),
+                               std::array<bool, 2>{true, true},
+                               false);
+
+    if constexpr (GraphViewType::is_multi_gpu) {
+      //  Create COO data structures
```

Comment:
@seunghwak - are there functions in `cugraph::detail` under the structure utilities that would be helpful here?

Reply:
Let me think more but I think

1) running adjacent_difference on the offsets array first,

2) call device_allgatherv using minor comm on tagged current vertices, counts (you get this by applying adjacent_difference on offsets) and indices

3) and converting counts to offsets

will work. Let me read this code more thoroughly.

Reply:
Yeah... so the entire code in the `if constexpr (GraphViewType::is_multi_gpu)` path is to construct a (tagged-current vertices, offsets, indices) array triplet for all the edges local to this GPU.

I think we should eventually hide this in our primitives, but currently you need to understand our 2D partitioning to see you actually don't need to collect array triplets from the entire set of GPUs.

You need to collect intersection indices for (current vertex, previous vertex) pairs only for the current vertices that can appear as edge sources.

For this purpose, you need to gather array triplets only from the GPUs in the same minor_comm.

This could be pretty confusing without proper understanding of our 2D partitioning scheme. You may speak with me or Joseph if you want to understand this more.

Anyways,

If you run thrust::adjacent_difference on `intersection_offsets`, you get `intersection_counts`.

Now you have (current vertex, previous vertex) pairs, intersection counts, and intersection indices.

You can run allgatherv on minor_comm.

Then you get aggregate (current vertex, previous vertex) pairs, aggregate intersection counts, and aggregate intersection indices.

And luckily, if (current_vertex, previous vertex) pairs are sorted in each GPU, aggregate (current vertex, previous vertex) pairs are also sorted.

Now you run thrust::exclusive|inclusive_scan on the aggregate intersection counts, you now have intersection offsets.